backend/services/auth_decorators.py

```diff
@@ -0,0 +1,67 @@
+from functools import wraps
+from fastapi import HTTPException, status
+from typing import Callable, Any
+import logging
+
+# Control what gets imported with 'from services.auth_decorators import *'
+__all__ = [
+    'require_admin',
+    'require_write',
+    'require_read'
+]
+
+logger = logging.getLogger(__name__)
+
+# Role hierarchy: admin > write > read
+REQUIRED_ROLE_HIERARCHY = {
+    'safot-admin': 3,
+    'safot-write': 2,
+    'safot-read': 1
+}
+
+def get_user_role_level(user_info: dict):
+    """Extract user role level from user_info"""
+    user_roles = user_info.get('roles', [])
+    for role in REQUIRED_ROLE_HIERARCHY:
+        if role in user_roles:
+            return REQUIRED_ROLE_HIERARCHY[role]
+    return 0
+
+def require_role(required_role: str):
+    """Decorator to require specific role for endpoint access"""
+    def decorator(func: Callable) -> Callable:
+        @wraps(func)
+        async def wrapper(*args, user_info: dict = None, **kwargs):
+            if not user_info:
+                raise HTTPException(
+                    status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                    detail="User info not available"
+                )           
+            
+            user_level = get_user_role_level(user_info)
+            required_level = REQUIRED_ROLE_HIERARCHY.get(required_role, 0)
+            if user_level < required_level:
```

Comment:
Again I would not assume all permissions are linear.
I would keep the decorators and in code here would add role specific logic that basically grants read and write for admin and grants read for write role when reading all user roles.